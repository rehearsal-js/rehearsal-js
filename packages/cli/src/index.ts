import { URL } from 'node:url';
import { resolve } from 'node:path';
import assert from 'node:assert';
import fs from 'node:fs/promises';
import { Command } from 'commander';
import { migrateCommand } from './commands/migrate/index.js';
import { upgradeCommand } from './commands/upgrade.js';
import type { PackageJson } from 'type-fest';

const __dirname = new URL('.', import.meta.url).pathname;
const { version } = JSON.parse(
  await fs.readFile(resolve(__dirname, '../package.json'), 'utf-8')
) as PackageJson;

const program = new Command();

assert(version, 'Has a rehearsal version');

program.name('rehearsal').version(version).addCommand(migrateCommand).addCommand(upgradeCommand);

export { program as rehearsal, migrateCommand, upgradeCommand };
