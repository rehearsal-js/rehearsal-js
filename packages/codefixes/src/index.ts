export { codefixes, glintCodeFixes, applyCodeFix, ContentDelegate } from './codefixes.js';
export { isInstallPackageCommand } from './typescript-codefix-collection.js';
export { hints } from './hints.js';
export { getDiagnosticOrder } from './get-diagnostics.js';
export * from './types.js';
