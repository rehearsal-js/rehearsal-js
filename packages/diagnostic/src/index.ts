export { default as diagnose } from './diagnose';
