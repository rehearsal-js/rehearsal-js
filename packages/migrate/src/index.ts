export { default as migrate } from './migrate';
