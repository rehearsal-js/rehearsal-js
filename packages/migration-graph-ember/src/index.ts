export { EmberAppPackage } from './entities/ember-app-package';
export { EmberAddonPackage } from './entities/ember-addon-package';

export {
  isApp as isEmberApp,
  isEngine as isEmberEngine,
  isAddon as isEmberAddon,
  getEmberAddonName,
} from './utils/ember';

export { discoverEmberPackages } from './utils/discover-ember-packages';

export {
  EmberAppPackageGraph,
  EmberAppPackageGraphOptions,
} from './entities/ember-app-package-graph';
export {
  EmberAddonPackageGraph,
  EmberAddonPackageGraphOptions,
} from './entities/ember-addon-package-graph';

export {
  EmberAppProjectGraph,
  EmberAppProjectGraphOptions,
} from './entities/ember-app-project-graph';
export {
  EmberAddonProjectGraph,
  EmberAddonProjectGraphOptions,
} from './entities/ember-addon-project-graph';
