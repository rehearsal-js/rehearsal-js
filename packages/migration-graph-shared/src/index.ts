import { resolve } from 'path';
import { readJsonSync } from 'fs-extra';

export function readPackageJson(pathToPackage: string): Record<string, any> {
  return readJsonSync(resolve(pathToPackage, 'package.json'));
}

export {
  Package,
  type PackageOptions,
  type PackageContainer,
  type PackageJson,
} from './entities/package';
export { formatter } from './utils/prettier';
export { isWorkspace } from './utils/workspace';
export { setNestedPropertyValue, removeNestedPropertyValue } from './utils/pojo';
export { isTesting, setupTestEnvironment } from './utils/environment';

export * from './graph';
export * from './types';
export { PackageGraph, PackageGraphOptions } from './entities/package-graph';
export { ProjectGraph, ProjectGraphOptions } from './entities/project-graph';

export { IPackage } from './entities/IPackage';
