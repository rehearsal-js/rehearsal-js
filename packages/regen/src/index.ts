export { regen, RegenInput } from './regen';
