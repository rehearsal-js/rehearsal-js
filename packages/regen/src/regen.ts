import { dirname, extname, resolve } from 'path';
import { PluginsRunner, RehearsalService } from '@rehearsal/service';
import {
  DiagnosticReportPlugin,
  LintPlugin,
  PrettierPlugin,
  ReRehearsePlugin,
} from '@rehearsal/plugins';
import ts from 'typescript';
import type { ListrContext } from 'listr2';
import type { Logger } from 'winston';
import type { Reporter } from '@rehearsal/reporter';
import type { ESLint } from 'eslint';

export type RegenInput = {
  basePath: string;
  entrypoint: string;
  sourceFiles: string[];
  tsConfigName?: string;
  eslintOptions?: ESLint.Options;
  reporter: Reporter;
  logger?: Logger;
  task?: ListrContext;
};

export type RegenOutput = {
  basePath: string;
  configFile: string;
  scannedFiles: Array<string>;
};

const { findConfigFile, parseJsonConfigFileContent, readConfigFile, sys } = ts;

export async function regen(input: RegenInput): Promise<RegenOutput> {
  const basePath = resolve(input.basePath);
  const tsConfigName = input.tsConfigName || 'tsconfig.json';
  const sourceFiles = input.sourceFiles || [resolve(basePath, 'index.ts')];
  const reporter = input.reporter;
  const logger = input.logger;

  //regen will only work on ts files
  const filteredSourceFiles = sourceFiles.filter((file) => extname(file) === '.ts');

  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
  const listrTask: { output: string } = input.task || { output: '' };

  logger?.debug('migration regen started');
  logger?.debug(`Base path: ${basePath}`);

  // eslint-disable-next-line @typescript-eslint/unbound-method
  const configFile = findConfigFile(basePath, sys.fileExists, tsConfigName);

  if (!configFile) {
    const message = `Config file '${tsConfigName}' not found in '${basePath}'`;
    throw Error(message);
  }

  logger?.debug(`config file: ${configFile}`);

  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
  const { config } = readConfigFile(configFile, (filepath: string, encoding?: string) =>
    sys.readFile(filepath, encoding)
  );

  const { options, fileNames: someFiles } = parseJsonConfigFileContent(
    config,
    sys,
    dirname(configFile),
    {},
    configFile
  );

  const fileNames = [...new Set([...someFiles, ...filteredSourceFiles])];

  logger?.debug(`fileNames: ${JSON.stringify(fileNames)}`);

  const commentTag = '@rehearsal';

  const service = new RehearsalService(options, fileNames);

  const runner = new PluginsRunner({ basePath, service, reporter, logger })
    .queue(new ReRehearsePlugin(), {
      commentTag,
    })
    .queue(new LintPlugin(), {
      eslintOptions: { cwd: basePath, useEslintrc: true, fix: true, ...input.eslintOptions },
      reportErrors: false,
    })
    .queue(new DiagnosticReportPlugin(), {
      commentTag,
    })
    .queue(new PrettierPlugin(), {})
    .queue(new LintPlugin(), {
      eslintOptions: { cwd: basePath, useEslintrc: true, fix: true, ...input.eslintOptions },
      reportErrors: true,
    });

  for await (const _ of runner.run(fileNames, { log: (message) => (listrTask.output = message) })) {
    // no ops for yield
  }

  reporter.saveCurrentRunToReport(basePath, input.entrypoint || '');

  return {
    basePath,
    configFile,
    scannedFiles: fileNames,
  };
}
