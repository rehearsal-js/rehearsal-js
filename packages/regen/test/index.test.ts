import { resolve } from 'path';
import { copyFileSync, mkdirSync, rmSync, readFileSync, existsSync } from 'fs';
import { fileURLToPath } from 'url';
import { Reporter } from '@rehearsal/reporter';
import { describe, test, beforeEach, afterEach, expect } from 'vitest';
import { outputFileSync } from 'fs-extra';
import { regen, RegenInput } from '../src';

const __dirname = fileURLToPath(new URL('.', import.meta.url));
const basePath = resolve(__dirname);
const expectedDir = resolve(basePath, 'fixtures', 'output');
const srcDir = resolve(basePath, 'fixtures', 'src');
const generatedDir = resolve(basePath, 'tmp');
const jsonReportPath = resolve(generatedDir, '.rehearsal-report.json');

let reporter: Reporter;
let regenInput: RegenInput;

describe('regen', () => {
  beforeEach(() => {
    cleanDir(generatedDir);
    mkdirSync(generatedDir);

    reporter = new Reporter({
      tsVersion: '',
      projectName: '@rehearsal/test',
      basePath,
      commandName: '@rehearsal/migrate',
    });

    regenInput = {
      basePath: resolve(basePath, 'tmp'),
      sourceFiles: [],
      reporter,
      entrypoint: '',
    };
  });

  afterEach(() => {
    cleanDir(generatedDir);
  });

  test('should scan files and generate report', async () => {
    prepareInputFiles(['test1.ts']);

    const { scannedFiles } = await regen(regenInput);
    expect(scannedFiles.length).toBe(1);

    reporter.saveReport(jsonReportPath);
    expect(existsSync(jsonReportPath)).toBeTruthy();
  });

  test('should scan ts files but ignore js files', async () => {
    prepareInputFiles(['test1.ts', 'test2.js']);

    const { scannedFiles } = await regen(regenInput);
    expect(scannedFiles.length).toBe(1);
    expect(scannedFiles[0].includes('test1.ts')).toBeTruthy();

    reporter.saveReport(jsonReportPath);
    const report = JSON.parse(readFileSync(jsonReportPath).toString());
    const { items } = report;
    expect(JSON.stringify(items, null, 2)).toMatchSnapshot();

    expect(fileOutputMatched('test1.ts')).toBeTruthy();
  });

  test('should generate correct rehearsal comments for a group of files', async () => {
    prepareInputFiles(['test1.ts', 'test3.ts']);

    const { scannedFiles } = await regen(regenInput);
    expect(scannedFiles.length).toBe(2);

    reporter.saveReport(jsonReportPath);

    const report = JSON.parse(readFileSync(jsonReportPath).toString());
    const { items } = report;
    expect(JSON.stringify(items, null, 2)).toMatchSnapshot();

    expect(fileOutputMatched('test3.ts')).toBeTruthy();
    expect(fileOutputMatched('test1.ts')).toBeTruthy();
  });

  test('should report lint errors', async () => {
    prepareInputFiles(['test4.ts']);

    const { scannedFiles } = await regen(regenInput);
    expect(scannedFiles.length).toBe(1);
    reporter.saveReport(jsonReportPath);

    const report = JSON.parse(readFileSync(jsonReportPath).toString());
    const { items } = report;
    expect(JSON.stringify(items, null, 2)).toMatchSnapshot();

    expect(fileOutputMatched('test4.ts')).toBeTruthy();
  });
});

function fileOutputMatched(filename: string): boolean {
  const generatedFile = resolve(generatedDir, filename);
  const generatedContent = readFileSync(generatedFile, 'utf-8');
  const expectedContent = readFileSync(`${expectedDir}/${filename}.output`, 'utf-8');
  return generatedContent === expectedContent;
}

function prepareInputFiles(srcFiles: string[]): string[] {
  createTSConfig(generatedDir);
  return srcFiles.map((file) => {
    const originalPath = resolve(srcDir, file);
    const generatedPath = resolve(generatedDir, file);
    copyFileSync(originalPath, generatedPath);
    return generatedPath;
  });
}

function createTSConfig(dir: string): void {
  const configStr = `{
    "extends": "../tsconfig",
    "include": ["."]
  }
  `;
  const configPath = resolve(dir, 'tsconfig.json');
  outputFileSync(configPath, configStr);
}

function cleanDir(dir: string): void {
  rmSync(dir, { recursive: true, force: true });
}
