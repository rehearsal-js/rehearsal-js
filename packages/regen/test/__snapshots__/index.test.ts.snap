// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`regen > should generate correct rehearsal comments for a group of files 1`] = `
"[
  {
    \\"analysisTarget\\": \\"tmp/test3.ts\\",
    \\"type\\": 0,
    \\"ruleId\\": \\"TS2345\\",
    \\"category\\": \\"Error\\",
    \\"message\\": \\"Argument of type 'string' is not assignable to parameter of type 'number'.\\",
    \\"hint\\": \\"Argument of type 'string' is not assignable to parameter of type 'number'. Consider verifying both types, using type assertion: '('5' as string)', or using type guard: 'if ('5' instanceof string) { ... }'.\\",
    \\"hintAdded\\": true,
    \\"nodeKind\\": \\"StringLiteral\\",
    \\"nodeText\\": \\"'5'\\",
    \\"helpUrl\\": \\"https://stackoverflow.com/questions/42421501/error-ts2345-argument-of-type-t-is-not-assignable-to-parameter-of-type-objec\\",
    \\"nodeLocation\\": {
      \\"startLine\\": 6,
      \\"startColumn\\": 5,
      \\"endLine\\": 6,
      \\"endColumn\\": 8
    }
  },
  {
    \\"analysisTarget\\": \\"tmp/test3.ts\\",
    \\"type\\": 0,
    \\"ruleId\\": \\"TS6133\\",
    \\"category\\": \\"Error\\",
    \\"message\\": \\"'baz' is declared but its value is never read.\\",
    \\"hint\\": \\"The variable 'baz' is never read or used. Remove the variable or use it.\\",
    \\"hintAdded\\": true,
    \\"nodeKind\\": \\"Identifier\\",
    \\"nodeText\\": \\"baz\\",
    \\"helpUrl\\": \\"https://stackoverflow.com/search?tab=votes&q=TS6133\\",
    \\"nodeLocation\\": {
      \\"startLine\\": 9,
      \\"startColumn\\": 5,
      \\"endLine\\": 9,
      \\"endColumn\\": 8
    }
  },
  {
    \\"analysisTarget\\": \\"tmp/test3.ts\\",
    \\"type\\": 0,
    \\"ruleId\\": \\"TS2322\\",
    \\"category\\": \\"Error\\",
    \\"message\\": \\"Type 'number' is not assignable to type 'string'.\\",
    \\"hint\\": \\"The variable 'baz' has type 'string', but 'number' is assigned. Please convert 'number' to 'string' or change variable's type.\\",
    \\"hintAdded\\": true,
    \\"nodeKind\\": \\"Identifier\\",
    \\"nodeText\\": \\"baz\\",
    \\"helpUrl\\": \\"https://stackoverflow.com/questions/72139358/ts2322-type-typeof-statusenum-is-not-assignable-to-type-statusenum\\",
    \\"nodeLocation\\": {
      \\"startLine\\": 11,
      \\"startColumn\\": 1,
      \\"endLine\\": 11,
      \\"endColumn\\": 4
    }
  },
  {
    \\"analysisTarget\\": \\"tmp/test3.ts\\",
    \\"type\\": 1,
    \\"ruleId\\": \\"no-unused-vars\\",
    \\"category\\": \\"Error\\",
    \\"message\\": \\"'baz' is assigned a value but never used.\\",
    \\"hint\\": \\"'baz' is assigned a value but never used.\\",
    \\"hintAdded\\": false,
    \\"nodeKind\\": \\"Identifier\\",
    \\"nodeText\\": \\"\\",
    \\"helpUrl\\": \\"\\",
    \\"nodeLocation\\": {
      \\"startLine\\": 11,
      \\"startColumn\\": 1,
      \\"endLine\\": 11,
      \\"endColumn\\": 4
    }
  },
  {
    \\"analysisTarget\\": \\"tmp/test3.ts\\",
    \\"type\\": 1,
    \\"ruleId\\": \\"@typescript-eslint/no-unused-vars\\",
    \\"category\\": \\"Error\\",
    \\"message\\": \\"'baz' is assigned a value but never used.\\",
    \\"hint\\": \\"'baz' is assigned a value but never used.\\",
    \\"hintAdded\\": false,
    \\"nodeKind\\": \\"Identifier\\",
    \\"nodeText\\": \\"\\",
    \\"helpUrl\\": \\"\\",
    \\"nodeLocation\\": {
      \\"startLine\\": 11,
      \\"startColumn\\": 1,
      \\"endLine\\": 11,
      \\"endColumn\\": 4
    }
  }
]"
`;

exports[`regen > should report lint errors 1`] = `
"[
  {
    \\"analysisTarget\\": \\"tmp/test4.ts\\",
    \\"type\\": 1,
    \\"ruleId\\": \\"@typescript-eslint/no-explicit-any\\",
    \\"category\\": \\"Error\\",
    \\"message\\": \\"Unexpected any. Specify a different type.\\",
    \\"hint\\": \\"Unexpected any. Specify a different type.\\",
    \\"hintAdded\\": false,
    \\"nodeKind\\": \\"TSAnyKeyword\\",
    \\"nodeText\\": \\"\\",
    \\"helpUrl\\": \\"\\",
    \\"nodeLocation\\": {
      \\"startLine\\": 1,
      \\"startColumn\\": 26,
      \\"endLine\\": 1,
      \\"endColumn\\": 29
    }
  },
  {
    \\"analysisTarget\\": \\"tmp/test4.ts\\",
    \\"type\\": 1,
    \\"ruleId\\": \\"no-undef\\",
    \\"category\\": \\"Error\\",
    \\"message\\": \\"'console' is not defined.\\",
    \\"hint\\": \\"'console' is not defined.\\",
    \\"hintAdded\\": false,
    \\"nodeKind\\": \\"Identifier\\",
    \\"nodeText\\": \\"\\",
    \\"helpUrl\\": \\"\\",
    \\"nodeLocation\\": {
      \\"startLine\\": 3,
      \\"startColumn\\": 5,
      \\"endLine\\": 3,
      \\"endColumn\\": 12
    }
  },
  {
    \\"analysisTarget\\": \\"tmp/test4.ts\\",
    \\"type\\": 1,
    \\"ruleId\\": \\"no-undef\\",
    \\"category\\": \\"Error\\",
    \\"message\\": \\"'console' is not defined.\\",
    \\"hint\\": \\"'console' is not defined.\\",
    \\"hintAdded\\": false,
    \\"nodeKind\\": \\"Identifier\\",
    \\"nodeText\\": \\"\\",
    \\"helpUrl\\": \\"\\",
    \\"nodeLocation\\": {
      \\"startLine\\": 5,
      \\"startColumn\\": 5,
      \\"endLine\\": 5,
      \\"endColumn\\": 12
    }
  },
  {
    \\"analysisTarget\\": \\"tmp/test4.ts\\",
    \\"type\\": 1,
    \\"ruleId\\": \\"no-undef\\",
    \\"category\\": \\"Error\\",
    \\"message\\": \\"'console' is not defined.\\",
    \\"hint\\": \\"'console' is not defined.\\",
    \\"hintAdded\\": false,
    \\"nodeKind\\": \\"Identifier\\",
    \\"nodeText\\": \\"\\",
    \\"helpUrl\\": \\"\\",
    \\"nodeLocation\\": {
      \\"startLine\\": 7,
      \\"startColumn\\": 5,
      \\"endLine\\": 7,
      \\"endColumn\\": 12
    }
  },
  {
    \\"analysisTarget\\": \\"tmp/test4.ts\\",
    \\"type\\": 1,
    \\"ruleId\\": \\"no-undef\\",
    \\"category\\": \\"Error\\",
    \\"message\\": \\"'console' is not defined.\\",
    \\"hint\\": \\"'console' is not defined.\\",
    \\"hintAdded\\": false,
    \\"nodeKind\\": \\"Identifier\\",
    \\"nodeText\\": \\"\\",
    \\"helpUrl\\": \\"\\",
    \\"nodeLocation\\": {
      \\"startLine\\": 9,
      \\"startColumn\\": 5,
      \\"endLine\\": 9,
      \\"endColumn\\": 12
    }
  },
  {
    \\"analysisTarget\\": \\"tmp/test4.ts\\",
    \\"type\\": 1,
    \\"ruleId\\": \\"no-undef\\",
    \\"category\\": \\"Error\\",
    \\"message\\": \\"'console' is not defined.\\",
    \\"hint\\": \\"'console' is not defined.\\",
    \\"hintAdded\\": false,
    \\"nodeKind\\": \\"Identifier\\",
    \\"nodeText\\": \\"\\",
    \\"helpUrl\\": \\"\\",
    \\"nodeLocation\\": {
      \\"startLine\\": 11,
      \\"startColumn\\": 5,
      \\"endLine\\": 11,
      \\"endColumn\\": 12
    }
  },
  {
    \\"analysisTarget\\": \\"tmp/test4.ts\\",
    \\"type\\": 1,
    \\"ruleId\\": \\"no-undef\\",
    \\"category\\": \\"Error\\",
    \\"message\\": \\"'console' is not defined.\\",
    \\"hint\\": \\"'console' is not defined.\\",
    \\"hintAdded\\": false,
    \\"nodeKind\\": \\"Identifier\\",
    \\"nodeText\\": \\"\\",
    \\"helpUrl\\": \\"\\",
    \\"nodeLocation\\": {
      \\"startLine\\": 13,
      \\"startColumn\\": 5,
      \\"endLine\\": 13,
      \\"endColumn\\": 12
    }
  }
]"
`;

exports[`regen > should scan ts files but ignore js files 1`] = `"[]"`;
